frontend/persist.lua

```diff
@@ -0,0 +1,139 @@
+    bitser = {
+        id = "Bitser serializer",
+
+        serialize = function(t, file)
+            local ok, str = pcall(bitser.dumps, t)
+            if not ok then
+                return nil, "cannot serialize " .. tostring(t)
+            end
+            return str
+        end,
+
+        deserialize = function(file)
+            local f, err, str
+            f, err = io.open(file, "rb")
+            if not f then
+                return nil, err
+            end
+            str, err = f:read("*a")
+            f:close()
+            if not str then
+                return nil, err
+            end
+            local ok, t = pcall(bitser.loads, str)
+            if not ok then
+                return nil, "malformed serialized data"
+            end
+            return t
+        end,
+    },
+    -- dump: human readable, pretty printed, fast enough for most user cases.
+    dump = {
+        id = "Dump serializer",
+
+        serialize = function(t, file, as_bytecode)
+            local content, err = dump(t)
+            if not content then
+                return nil, string.format("cannot serialize table %s: %s", t, err)
+            end
+            local str
+            if as_bytecode then
+                str, err =  load("return " .. content)
+                if not str then
+                    print("cannot convert table to bytecode: %s, ignoring", err)
+                else
+                    str = string.dump(str, true)
+                end
+            end
+
+            if not str then
+                str = "return " .. content
+            end
+            return str
+        end,
+
+        deserialize = function(file)
+            local ok, t, err = pcall(dofile, file)
+            if not ok then
+                return nil, err
+            end
+            return t
+        end,
+    }
+}
+
+local Persist = {}
+
+function Persist:new(o)
+    o = o or {}
+    setmetatable(o, self)
+    self.__index = self
+    return o:init(o.path, o.codec)
+end
+
+function Persist:init(path, codec)
+    if type(path) ~= "string" then
+        return nil, "path is required"
+    end
+    self.path = path
+    self.codec = codec or "dump"
+    return self
+end
+
+function Persist:exists()
+    local mode = lfs.attributes(self.path, "mode")
+    if mode then
+        return mode == "file"
+    end
+end
+
+function Persist:timestamp()
+    return lfs.attributes(self.path, "modification")
+end
+
+function Persist:size()
+    return lfs.attributes(self.path, "size")
+end
+
+function Persist:stats()
+    local stats = string.format("file: %s, codec: %s, exists: %s, size: %d, timestamp: %d",
+        self.path, self.codec, self:exists(), self:size(), self:timestamp())
+    print(stats)
+end
+
+function Persist:loadFile()
+    local t, err = codecs[self.codec]:deserialize(self.path)
+    if not t then
+        return nil, err
+    end
+    return t
+end
+
+function Persist:saveFile(t, as_bytecode)
```

Comment:
I was under the impression that `load` was a reserved word in lua, because it was linted when used as a method, so I appended `File`. It isn't the case (still it is a [thing](http://www.lua.org/manual/5.1/manual.html#pdf-load) but works fine as a method.